### DISCUSSION TITLE
Type of loader/action is not inferred correctly when using redirect

### DISCUSSION THREAD

**Discussion by Marcdj-02** [remix-run/remix#6472] [Q&A] [answered]
With the following loader function:

```typescript
type LoaderData = {
  content: string;
}

export async function loader({ params, request }: LoaderFunctionArgs) {
  if (hasNoAccess) {
    return redirect('/login');
  }

  return json<LoaderData>({ content: 'Hello World' });
}
```

I would expect the return type of the loader to be:
```typescript
Promise<TypedResponse<LoaderData>>
```
But in this case it is:
```typescript
Promise<Response>
```

This means that the `useLoaderData<typeof loader>` cannot infer the types of the loaderData.

Anyone got any suggestions?

**Comment by machour** [comment]
Hey,

First thing first, you shouldn't need to define/use `LoaderData`, Remix will be able to infer that for you.

~~Your loader is currently returning either a redirection, or a json object, so it would be wrong for inference to lie to you and say you only get a `LoaderData` from your loader.~~

What you can do instead, is change `return redirect('/login')` to `throw redirect('/login')` which will do the same thing. This way, your loader only returns the struct, and inference will work as expected:

```tsx
export async function loader({ params, request }: LoaderFunctionArgs) {
  if (hasNoAccess) {
    throw redirect('/login');
  }

  return json({ content: 'Hello World' });
}
```

**Reply by Marcdj-02** [reply]
First of all. Thanks for replying.

Regarding your first point, the types are not inferred right if I don't specify LoaderData in the following case:
```typescript
export async function loader({ params, request }: LoaderFunctionArgs) {
  if(hasNoAccess) {
    return json({});
  }

  return json({ content: 'Hello world' });
}
```

This will result in a return type of
```typescript
Promise<TypedResponse<{}>>
```
Instead I expect: 
```typescript
Promise<TypedResponse<{ content?: string }>>
```

Secondly, I am aware of the `throw redirect()` instead of `return redirect()` option, but this calls for the developer to be aware of this (If not the loaderData is just given a type of 'any'). Maybe we could make it more intuitive by changing the return type of the loader which would then reflect that it could be a json response or a redirect response. This then can be caught in the useLoaderData's type inference?

For example: 
```typescript
Promise<TypedResponse<LoaderData> | RedirectResponse>
```

Then the useLoaderData hook can just use the TypedResponse and disregard the RedirectResponse

**Reply by machour** [reply, marked as answer]
What remix version are you using?

I'm using 1.16.0 and seeing this behavior:

<img width="669" alt="Screenshot 2023-05-24 at 1 57 59 PM" src="https://github.com/remix-run/remix/assets/304450/50db224c-9bbc-4f12-b576-af77d3d16f39">

✅  inference of with json() & redirect()

<img width="526" alt="Screenshot 2023-05-24 at 1 59 10 PM" src="https://github.com/remix-run/remix/assets/304450/d9587b20-4b44-4108-95e6-bbeda0d21e95">

✅  inference of with 2 non-empty json()

<img width="538" alt="Screenshot 2023-05-24 at 1 58 28 PM" src="https://github.com/remix-run/remix/assets/304450/7c587810-a327-4a03-ac56-9e3169e3dfff">

🐛  inference is broken if you return an empty object 


Also, could you possibly be mixing imports and importing `json` or `useLoaderData` from `react-router` instead of `@remix-run/xxx` ?

**Reply by machour** [reply]
Two side notes:

- https://github.com/remix-run/remix/pull/6449 (available in v0.0.0-nightly-d56574b-20230523) makes the whole thing much more readable 🥰 

<img width="428" alt="Screenshot 2023-05-24 at 2 05 14 PM" src="https://github.com/remix-run/remix/assets/304450/19d0406a-6c4d-4cc1-b364-da71e0adb560">

but returning `json({})` in a branch still makes the whole inference trip & fail:

<img width="427" alt="Screenshot 2023-05-24 at 2 06 10 PM" src="https://github.com/remix-run/remix/assets/304450/399ed4be-3841-491b-87f5-4a1025b20ecc">

We may want to open a bug report for this

/cc @pcattori 

- You may be interested in checking https://github.com/kiliman/remix-typedjson

**Reply by kiliman** [reply]
Yeah, `remix-typedjson` kinds of punts on type inference by simply using the type that TypeScript sees since I don't have to worry about the JSON representation. My library simply converts the JSON payload back into the original object (with the correct types).

**Reply by Marcdj-02** [reply]
Yeah, you're right. The import of redirect was from the wrong package. Thanks for your tip. 

Simplified types will be much better indeed. Plus, the fix of the empty object would be nice.

**Reply by iamkd** [reply]
Hi! Sorry for a tangential question, but I've been doing exactly this right now - replacing `return` with `throw` to improve types across the codebase. Because of this I've been a bit confused about the described behavior of `ErrorBoundary` v2. Is `redirect` exempt from being caught by ErrorBoundary? 

I'm testing it right now and it seems that any thrown Response (doesn't matter if it's 2xx or 4xx) is caught by it, but `redirect` or a manual 302 Response is working as expected. I'd love to have a more clear explanation/documentation on any possible exceptions, thanks!

Edit:
Okay, after putting my thoughts into words I think I get the reason. The response headers/status are not modified by the error boundary, the boundary is just the response body. So... 302 still stays 302 when it reaches the browser and the redirect works as expected. My apologies to everyone who got the notifications 😄

**Reply by pcattori** [reply]
<img width="335" alt="Screenshot 2023-05-24 at 10 06 41 AM" src="https://github.com/remix-run/remix/assets/1477317/41bd9534-0a01-44e6-80a9-ef08422a7476">

Types collapsing when conditionally returning empty object is not a Remix issue. Reproduced in pure TS.

[Link to Typescript playground](https://www.typescriptlang.org/play?#code/DYUwLgBAlgJhC8EA8AVANAPgBQA8BcEKAlAhhDgFAWiQBmCEWJ8ZA3hRJ9PVgLICGYABYA6AE78AdjAD2AWyYQyABhEBWEuy5cx4AK5jJ0GFlYQhIYMBkEARAHcZY4DFsQAvkQ5d33zrrADI1hTTwpfCjAATwAHEAgAJQYE-UMUWJAkaLiZeloMCgB6Qq4APQB+IA)

**Reply by bacharSalleh** [reply]
any progress on this issue?. it is really ruin the types. I think the only solution is type casting 😢 for now

**Reply by Marcdj-02** [reply]
> any progress on this issue?. it is really ruin the types. I think the only solution is type casting 😢 for now

Are you importing from the correct package?

**Reply by kiliman** [reply]
@bacharSalleh I recommend you create a new discussion with code so we can help you better.
